File: apps/automate-go/internal/processor/graph.go
package processor

import (
	"tdata/automate/internal/db"
)

type Graph struct {
	StartNode *db.AutomationFlowNode
	Routes    map[string]map[string]*db.AutomationFlowNode // nodeID -> (edgeName -> nextNode)
}

// generateGraph builds the graph from nodes and edges arrays.
func BuildGraph(flow *db.AutomationFlow) *Graph {
	nodes := flow.Nodes
	edges := flow.Edges

	// Map to store nodes by their IDs for constant-time lookups
	nodeMap := make(map[string]*db.AutomationFlowNode)

	// Initialize the graph structure
	g := &Graph{
		Routes: make(map[string]map[string]*db.AutomationFlowNode),
	}

	// Populate the node map for fast access by node ID
	for _, node := range nodes {
		if node.Type == db.TriggerNode {
			g.StartNode = &node
		}
		nodeMap[node.ID] = &node
	}

	// Build graph from edges
	for _, edge := range edges {
		sourceNode := nodeMap[edge.Source]
		targetNode := nodeMap[edge.Target]

		edgeName := "true"

		// If the source or target node is nil, skip this edge (invalid data)
		if sourceNode == nil || targetNode == nil {
			continue
		}

		// If it's a PlaceholderNode, don't add it as a valid next node
		if targetNode.Type == db.PlaceholderNode {
			targetNode = nil
		}

		if edge.SourceHandle != nil {
			edgeName = *edge.SourceHandle
		}

		// Create the map for the source node if it doesn't exist
		if _, ok := g.Routes[edge.Source]; !ok {
			g.Routes[edge.Source] = make(map[string]*db.AutomationFlowNode)
		}

		// Add the edge to the graph
		g.Routes[edge.Source][edgeName] = targetNode
	}

	return g
}

// GetNextNode retrieves the next node based on the current node ID and the sourceHandle.
// It returns nil if the next node is a PlaceholderNode (leaf node) or doesn't exist.
func (g *Graph) GetNextNode(currentNodeID, sourceHandle string) *db.AutomationFlowNode {
	// Find the edge with the matching sourceHandle for the current node
	if nextNode, ok := g.Routes[currentNodeID][sourceHandle]; ok {
		return nextNode
	}
	return nil
}

File: apps/automate/src/processor/graph.ts
import { FlowNode, FlowEdge, NodeType, NodeDataMap, AutomationFlow } from "@tdata/shared/types";

export class FlowGraph {
  startNode: FlowNode = {} as FlowNode;
  private nodeMap: Map<
    string,
    FlowNode & {
      edges: Record<string, string>; // branch name to target node id
    }
  > = new Map();

  constructor(flow: AutomationFlow) {
    this.buildGraph(flow.nodes, flow.edges);
  }

  private buildGraph(flowNodes: FlowNode[], flowEdges: FlowEdge[]) {
    flowNodes.forEach((node) => {
      if (node.type == "TriggerNode") {
        this.startNode = node;
      }
      this.nodeMap.set(node.id, {
        ...node,
        edges: {},
      });
    });

    // Then, add edges
    flowEdges.forEach((edge) => {
      const sourceNode = this.nodeMap.get(edge.source);
      if (sourceNode) {
        // Use sourceHandle or default to 'true'
        const branchName = this.determineBranchName(sourceNode.type, edge.sourceHandle || undefined);
        sourceNode.edges[branchName] = edge.target;
      }
    });
  }

  private determineBranchName(nodeType: NodeType, sourceHandle?: string): string {
    // For condition nodes, use sourceHandle
    if (nodeType === "ConditionNode") {
      return sourceHandle || "false";
    }
    // For all other nodes, use 'true'
    return "true";
  }

  // O(1) get next node
  getNextNode(currentNodeId: string, branch: string = "true"): FlowNode | null {
    const node = this.nodeMap.get(currentNodeId);
    const nextNodeId = node ? node.edges[branch] || null : null;
    return nextNodeId ? this.nodeMap.get(nextNodeId) || null : null;
  }

  // O(1) get node data
  getNodeData<T extends NodeType>(nodeId: string): NodeDataMap[T] | null {
    const node = this.nodeMap.get(nodeId);
    return node ? (node.data as NodeDataMap[T]) : null;
  }

  // O(1) get node type
  getNodeType(nodeId: string): NodeType | null {
    const node = this.nodeMap.get(nodeId);
    return node ? node.type : null;
  }
}
